### PR TITLE
Move ObjC call APIs into a separate module in prep for the Metal backend

### PR DESCRIPTION
**Pull request by k-ye**

### CODE REVIEW

File: taichi/platform/mac/objc_api.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+using nsobj_unique_ptr = std::unique_ptr<O, NsObjDeleter<O>>;
+
+template <typename O>
+nsobj_unique_ptr<O> wrap_as_nsobj_unique_ptr(O *nsobj) {
+  return nsobj_unique_ptr<O>(nsobj);
+}
+
+// Prepend "TC_" to native ObjC type names, otherwise clang-format thinks this
+// is an ObjC file and is not happy formatting it.
+struct TC_NSString;
```
Comment (k-ye) [inline taichi/platform/mac/objc_api.h]: Halide defined these incomplete struct types to have type safety (https://github.com/halide/Halide/blob/60b039df975d2978240f58b1fbfbdd4b1e97ec09/src/runtime/metal.cpp#L23). I followed that design here.

**Comment by k-ye** [comment]
Hey, I'm out most of the day and will address any comment later tonight or so.

**Comment by yuanming-hu** [comment]
Thanks very much and no rush! I have plenty of tasks in my queue and I'll take a pass tonight.

**Comment by k-ye** [comment]
Sorry about breaking non-Mac platforms. X(

**Comment by yuanming-hu** [comment]
No worries - working on fixing AppVeyor for windows tests...

**Pull request merged**